folio-spring-i18n/src/main/java/org/folio/spring/model/TranslationMap.java

```diff
@@ -0,0 +1,190 @@
+   * existing TranslationMaps for use as fallbacks for another (for caching reasons, this
+   * makes more sense than re-reading files each time)
+   *
+   * @param newLocale the locale for the new TranslationMap to use
+   * @return the TranslationMap with {@code newLocale}
+   */
+  public TranslationMap withLocale(Locale newLocale) {
+    return new TranslationMap(
+      newLocale,
+      this.file,
+      this.patterns,
+      this.fallback
+    );
+  }
+
+  /**
+   * Get the ICU format string associated with a given key for this translation.
+   *
+   * @param key the key to lookup
+   * @return one of the following, with this precedence:
+   *   <ol>
+   *     <li>the ICU format string for this locale, if it is known</li>
+   *     <li>the ICU format string for a locale with a worse {@code TranslationMatchQuality},
+   *       such as one without a specific country, recursing through the {@code fallback}</li>
+   *     <li>the ICU format string for the server's default locale</li>
+   *     <li>the provided key</li>
+   *   </ol>
+   */
+  public String get(String key) {
+    String pattern = this.patterns.get(key);
+
+    if (pattern == null && this.fallback != null) {
+      pattern = this.fallback.get(key);
+    }
+
+    if (pattern == null) {
+      log.error("Could not resolve key {} in any translation", key);
+      // fallback to translation key itself
+      pattern = key;
+    }
+
+    return pattern;
+  }
+
+  /**
+   * Format an ICU format string (found by its key), supplying a series of named arguments as key
+   * value pairs.  For example: {@code format("Hello {name}", "name", parameterValue)}
+   *
+   * @param key the key of the format string
+   * @param args pairs of keys and values to interpolate
+   * @return the formatted string
+   */
+  public String format(String key, Object... args) {
+    for (int i = 0; i < args.length; i++) {
+      // Convert LocalDate to Date
+      // Sadly, ICU formatting strings only support date formats with the old Date class :(
+      if (args[i] instanceof LocalDate date) {
+        args[i] =
+          Date.from(date.atStartOfDay(ZoneId.systemDefault()).toInstant());
+      }
+      // Same for LocalTime
+      if (args[i] instanceof LocalTime time) {
+        args[i] =
+          Date.from(
+            time
+              .atDate(LocalDate.now())
+              .atZone(ZoneId.systemDefault())
+              .toInstant()
+          );
+      }
+    }
+    return MessageFormat.format(this.get(key), buildMap(args));
+  }
+
+  /**
+   * Get a Map&lt;String, Object&gt; from a set of String -&gt; Object pairs.
+   *
+   * @param args pairs of elements, e.g. key1, value1, key2, value2, ...
+   * @return the map of each key =&gt; value
+   * @throws IllegalArgumentException if an odd number of parameters is passed
+   */
+  private static Map<String, Object> buildMap(Object... args) {
+    Map<String, Object> map = new HashMap<>();
```

Comment:
Tiniest of nits: I'd move this declaration until after the input validation, so that the declaration happens right next to where it's used